Revise ViewTrait deprecations for legacy View interop

ndellingwood opened the pull request:
Changes made to restore legacy View support with Sacado/Stokhos in Trilinos

ndellingwood committed:
Revise ViewTrait deprecations for legacy View interop

Changes made to restore legacy View support with Sacado/Stokhos in Trilinos

Signed-off-by: Nathan Ellingwood <ndellin@sandia.gov>

ndellingwood committed:
apply clang-format

Signed-off-by: Nathan Ellingwood <ndellin@sandia.gov>

crtrott reviewed (changes requested):
When legacy view is on that file is basically unused, how does this do anything?

ndellingwood commented:
> When legacy view is on that file is basically unused, how does this do anything?

@crtrott Which file are you referring to?
Sacado and Stokhos are missing some of the aliases from ViewDataAnalysis, e.g. `data_type`. I tried introducing them as aliases to `type` etc. (in Sacado/Stokhos), but still ran into compilation issues. I reintroduced some of the changes/removals from #7360 until I got compilation working, I can try paring some of them back here

ndellingwood commented:
I tested individually reverting changes made to each file here:
1. Revert of changes to `core/src/View/Kokkos_ViewLegacy.hpp` hits compilation error:
```
/Users/ndellin/Research/trilinos/Trilinos-kp/packages/stokhos/test/UnitTest/Stokhos_KokkosViewMPVectorUnitTest.hpp:203:20: error: call of overloaded 'deep_copy(ViewType&, Scalar&)' is       ambiguous
  203 |   Kokkos::deep_copy( v, val );
```
2. Revert of changes to `core/src/View/Kokkos_ViewTraits.hpp` hits compilation error:
```
/Users/ndellin/Research/trilinos/Trilinos-kp/packages/stokhos/test/UnitTest/Stokhos_KokkosViewMPVectorUnitTest_Serial.cpp:20:1:   required from here
/Users/ndellin/Research/trilinos/Trilinos-kp/kokkos/core/src/View/Kokkos_ViewTraits.hpp:491:9: error: no type named 'data_type' in 'using Kokkos::ViewTraits<Sacado::MP::Vector<Stokhos::     StaticFixedStorage<int, double, 16, Kokkos::Serial> >*, Kokkos::Serial>::data_analysis = struct Kokkos::Impl::ViewDataAnalysis<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double,    16, Kokkos::Serial> >*, Kokkos::LayoutRight, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> > >' {aka 'struct Kokkos::Impl::ViewDataAnalysis<Sacado::MP::    Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> >*, Kokkos::LayoutRight, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> > >'}
  491 |   using data_type           = typename data_analysis::data_type;
      |         ^~~~~~~~~                        
/Users/ndellin/Research/trilinos/Trilinos-kp/kokkos/core/src/View/Kokkos_ViewTraits.hpp:492:9: error: no type named 'const_data_type' in 'using Kokkos::ViewTraits<Sacado::MP::               Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> >*, Kokkos::Serial>::data_analysis = struct Kokkos::Impl::ViewDataAnalysis<Sacado::MP::Vector<Stokhos::                   StaticFixedStorage<int, double, 16, Kokkos::Serial> >*, Kokkos::LayoutRight, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> > >' {aka 'struct Kokkos::Impl:: ViewDataAnalysis<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> >*, Kokkos::LayoutRight, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16,     Kokkos::Serial> > >'}
  492 |   using const_data_type     = typename data_analysis::const_data_type;
      |         ^~~~~~~~~~~~~~~                                                 
/Users/ndellin/Research/trilinos/Trilinos-kp/kokkos/core/src/View/Kokkos_ViewTraits.hpp:493:9: error: no type named 'non_const_data_type' in 'using Kokkos::ViewTraits<Sacado::MP::           Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> >*, Kokkos::Serial>::data_analysis = struct Kokkos::Impl::ViewDataAnalysis<Sacado::MP::Vector<Stokhos::                   StaticFixedStorage<int, double, 16, Kokkos::Serial> >*, Kokkos::LayoutRight, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> > >' {aka 'struct Kokkos::Impl:: ViewDataAnalysis<Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16, Kokkos::Serial> >*, Kokkos::LayoutRight, Sacado::MP::Vector<Stokhos::StaticFixedStorage<int, double, 16,     Kokkos::Serial> > >'}
  493 |   using non_const_data_type = typename data_analysis::non_const_data_type; 
```
* Introducing aliases e.g.  `using data_type = type` in Sacado and Stokhos' ViewDataAnalysis specs would help bypass this issue, but left issue 1. above unresolved
3. Revert of changes to `core/src/Kokkos_View.hpp` had no impact, I'll commit the revert those changes

ndellingwood committed:
revert changes to core/src/Kokkos_View.hpp

unused for legacy View

Signed-off-by: Nathan Ellingwood <ndellin@sandia.gov>

ndellingwood commented:
@crtrott I reverted the changes to `core/src/Kokkos_View.hpp` [6c952bb](https://github.com/kokkos/kokkos/pull/8614/commits/6c952bb0a0f32e445d0c084de5fbcb033b090668) , could you re-review please?

ndellingwood commented:
jenkins CI errors seem to be unrelated to changes in the PR?

* [SYCL-OneAPI](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8614/2/pipeline-overview/): issues with the container
```
Step 13/15 : RUN curl -LOJ "https://developer.codeplay.com/api/v1/products/download?product=oneapi&variant=nvidia&version=2025.0.0&filters[]=12.0&filters[]=linux" &&     chmod +x oneapi-for-nvidia-gpus-2025.0.0-cuda-12.0-linux.sh &&     ./oneapi-for-nvidia-gpus-2025.0.0-cuda-12.0-linux.sh

 ---> Running in 0d0f1da95629

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current

                           
      Dload  Upload   Total   Spent    Left  Speed


  0     0    0     0   
 0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  9379  100  9379    0     0  27385      0 --:--:-- --:--:-- --:--:-- 27344

chmod: cannot access 'oneapi-for-nvidia-gpus-2025.0.0-cuda-12.0-linux.sh': No such file or directory
```

* [CUDA-12.5.1-Clang-17-RDC](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8614/2/pipeline-overview/?selected-node=392): Linker issues?
```
[CUDA-12.5.1-Clang-17-RDC] /usr/bin/ld: ../src/libkokkoscore.a(Kokkos_Cuda_Instance.cpp.o): in function `Kokkos::Impl::CudaInternal::initialize(CUstream_st*)':

[CUDA-12.5.1-Clang-17-RDC] /var/jenkins/workspace/Kokkos_PR-8614/core/src/Cuda/Kokkos_Cuda_Instance.cpp:272:(.text+0x10cd): undefined reference to `cuStreamGetCtx'

[CUDA-12.5.1-Clang-17-RDC] /usr/bin/ld: /var/jenkins/workspace/Kokkos_PR-8614/core/src/Cuda/Kokkos_Cuda_Instance.cpp:273:(.text+0x10df): undefined reference to `cuCtxPushCurrent_v2'

[CUDA-12.5.1-Clang-17-RDC] /usr/bin/ld: /var/jenkins/workspace/Kokkos_PR-8614/core/src/Cuda/Kokkos_Cuda_Instance.cpp:274:(.text+0x10ef): undefined reference to `cuCtxGetDevice'

[CUDA-12.5.1-Clang-17-RDC] /usr/lib/llvm-17/bin/clang-linker-wrapper: error: 'ld' failed

[CUDA-12.5.1-Clang-17-RDC] clang++-17: error: linker command failed with exit code 1 (use -v to see invocation)

[CUDA-12.5.1-Clang-17-RDC] make[2]: *** [core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_DeviceAndThreads.dir/build.make:100: core/unit_test/Kokkos_CoreUnitTest_DeviceAndThreads] Error 1

[CUDA-12.5.1-Clang-17-RDC] make[1]: *** [CMakeFiles/Makefile2:2638: core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_DeviceAndThreads.dir/all] Error 2

[CUDA-12.5.1-Clang-17-RDC] make[1]: *** Waiting for unfinished jobs....
```

* [CUDA-13.0-NVCC-DEBUG](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8614/2/pipeline-overview/?selected-node=267): Similar to the job above...
```
[CUDA-13.0-NVCC-DEBUG] /usr/bin/ld: ../src/libkokkoscore.so.5.0.99: undefined reference to `cuStreamGetCtx'

[CUDA-13.0-NVCC-DEBUG] /usr/bin/ld: ../src/libkokkoscore.so.5.0.99: undefined reference to `cuCtxGetDevice'

[CUDA-13.0-NVCC-DEBUG] /usr/bin/ld: ../src/libkokkoscore.so.5.0.99: undefined reference to `cuCtxPushCurrent_v2'

[CUDA-13.0-NVCC-DEBUG] collect2: error: ld returned 1 exit status

[CUDA-13.0-NVCC-DEBUG] make[2]: *** [core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_DeviceAndThreads.dir/build.make:101: core/unit_test/Kokkos_CoreUnitTest_DeviceAndThreads] Error 1

[CUDA-13.0-NVCC-DEBUG] make[1]: *** [CMakeFiles/Makefile2:2419: core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_DeviceAndThreads.dir/all] Error 2

[CUDA-13.0-NVCC-DEBUG] make[1]: *** Waiting for unfinished jobs....
```

masterleinad commented:
> [SYCL-OneAPI](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8614/2/pipeline-overview/): issues with the container

Looks like the plugins are now part of `intel/llvm`, see https://github.com/intel/llvm/tree/sycl/unified-runtime/source/adapters.

science-enthusiast commented on core/src/View/Kokkos_ViewTraits.hpp:
```diff
@@ -488,20 +488,36 @@ struct ViewTraits {
+      typename data_analysis::const_scalar_array_type;
+  using non_const_scalar_array_type KOKKOS_DEPRECATED_WITH_COMMENT(
+      "Use non_const_data_type instead.") =
+      typename data_analysis::non_const_scalar_array_type;
```
I was one of the reviewers of #7360. I would like to understand this better. This part of the code can be reached only if both `KOKKOS_ENABLE_DEPRECATED_CODE_5` and `KOKKOS_ENABLE_IMPL_VIEW_LEGACY` are defined. In the other two places only `KOKKOS_ENABLE_IMPL_VIEW_LEGACY` needs to be defined.

If we can avoid constructs like `using scalar_array_type KOKKOS_DEPRECATED_WITH_COMMENT("Use data_type instead.") = typename data_analysis::scalar_array_type;` where there is a mismatch between what is in the comment and what is in the RHS, it is better.

science-enthusiast replied:
Before #7360 also, in `ViewDataAnalysis`, `scalar_array_type` was an alias to `type` and so on. Can't we just say `using scalar_array_type KOKKOS_DEPRECATED_WITH_COMMENT("Use data_type instead.") = data_type;` and so on, here? Since we do declare the following: https://github.com/kokkos/kokkos/pull/8614/files#diff-47ca725e5d6f25605453c8a6c24d408a510e6bc86a61c55b70caa2c8312fbaa9R491-R495

ndellingwood replied:
@science-enthusiast I initially thought that should work but ran into compilation issues. The reason I reintroduced the previous `scalar_array_type` alias (guarded by KOKKOS_ENABLE_IMPL_VIEW_LEGACY)  is that in Sacado and Stokhos they have View customizations, including specializations of `ViewDataAnalysis` that apparently handle the `scalar_array_type` alias differently than the Kokkos implementation; I tried for example to alias `data_type` in terms of `scalar_array_type` etc. in their ViewDataAnalysis specs, but that caused additional new compilation issues. Reintroducing `scalar_array_type` aliased from `data_analysis::scalar_array_type` (in Kokkos) recovered the previous behavior when using the legacy View

ndellingwood replied:
@science-enthusiast  I'll fix the copy-paste mismatch in the deprecated message comment , thanks for pointing that out

ndellingwood replied:
@science-enthusiast in case useful, here is a summary of the issues that occur if reverting the changes in the PR https://github.com/kokkos/kokkos/pull/8614#issuecomment-3457175991

science-enthusiast replied:
Thank you for the clarifications, @ndellingwood :+1:

ndellingwood commented:
@science-enthusiast re: the deprecation messages for the `scalar_array_type` alias etc., does something like this sound reasonable?
```
"This usage is only supported with KOKKOS_ENABLE_IMPL_VIEW_LEGACY enabled, which will be removed after the 5.0 release."
```

science-enthusiast commented:
> @science-enthusiast re: the deprecation messages for the `scalar_array_type` alias etc., does something like this sound reasonable?
> 
> ```
> "This usage is only supported with KOKKOS_ENABLE_IMPL_VIEW_LEGACY enabled, which will be removed after the 5.0 release."
> ```

Sure, @ndellingwood :+1: You can shorted the message like "Only supported with KOKKOS_ENABLE_IMPL_VIEW_LEGACY, to be removed after 5.0 release." if it will make it easier for you.

ndellingwood committed:
fix deprecation comment for scalar_array_type with legacy View

Signed-off-by: Nathan Ellingwood <ndellin@sandia.gov>

janciesko commented on core/src/View/Kokkos_ViewLegacy.hpp:
```diff
@@ -236,13 +236,23 @@ class View : public ViewTraits<DataType, Properties...> {
+           typename traits::memory_traits>,
+      View<typename traits::scalar_array_type, typename traits::array_layout,
+           typename traits::device_type, typename traits::memory_traits>>;
+#else
```
This change shouldn't be needed if `traits::data_type` resolves to `using data_type = typename data_analysis::type;` instead of `using data_type = typename data_analysis::data_type;` if `KOKKOS_ENABLE_IMPL_VIEW_LEGACY` is set. See your change in ViewTraits.hpp:L492

ndellingwood replied:
@janciesko Stokhos does not have a `data_type` alias in the ViewDataAnalysis specialization; I can introduce this (e.g. using data_type = type) there, but that assumes the pairing is correct for that impl

ndellingwood replied:
Dropping the changes results in compilation errors in Stokhos
```
/Users/ndellin/Research/trilinos/Trilinos-kp/packages/stokhos/test/UnitTest/Stokhos_KokkosViewMPVectorUnitTest.hpp:203:20: error: call of overloaded 'deep_copy(ViewType&, Scalar&)' is       ambiguous
  203 |   Kokkos::deep_copy( v, val );
```

janciesko replied:
hm, ok

janciesko commented on core/src/View/Kokkos_ViewTraits.hpp:
```diff
@@ -488,20 +488,37 @@ struct ViewTraits {
 #ifdef KOKKOS_ENABLE_DEPRECATED_CODE_5
   //------------------------------------
   // Compatible array of trivial type traits:
+#ifdef KOKKOS_ENABLE_IMPL_VIEW_LEGACY
```
Probably move this outside of  the  `KOKKOS_ENABLE_DEPRECATED_CODE_5` block as currently `KOKKOS_ENABLE_DEPRECATED_CODE_5` and `KOKKOS_ENABLE_IMPL_VIEW_LEGACY` are orthogonal
.

ndellingwood replied:
I agree they are orthogonal, but I'm not sure how to avoid the nesting of the if-guards to prevent duplication of alias definitions

crtrott commented:
What I meant at some point is that the LegacyView file is only included if Legacy View is enabled and vice versa. So in that file the define is always gonna be there. No reason to have else branches or ifdef protection at all...

ndellingwood commented:
> What I meant at some point is that the LegacyView file is only included if Legacy View is enabled and vice versa. So in that file the define is always gonna be there. No reason to have else branches or ifdef protection at all...

@crtrott okay, that makes sense now, thanks. I'll push up that change to the LegacyView file

ndellingwood committed:
drop KOKKOS_ENABLE_IMPL_VIEW_LEGACY guards in Kokkos_ViewLegacy

unnecessary since Kokkos_ViewLegacy.hpp only included when
KOKKOS_ENABLE_IMPL_VIEW_LEGACY is true

Signed-off-by: Nathan Ellingwood <ndellin@sandia.gov>

ndellingwood commented:
@crtrott updated and rebased on develop to include the recent CI fixes

ndellingwood commented:
Jenkins failures are system-related, not due to changes in the PR

CUDA-12.2-NVHPC-AS-HOST-COMPILER
```
[CUDA-12.2-NVHPC-AS-HOST-COMPILER] [  4%] Linking CXX executable Kokkos_CoreUnitTest_DeviceAndThreads
[CUDA-12.2-NVHPC-AS-HOST-COMPILER] nvlink fatal   : Could not open input file '/usr/lib/x86_64-linux-gnu/libcuda.so'
[CUDA-12.2-NVHPC-AS-HOST-COMPILER] make[2]: *** [core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_DeviceAndThreads.dir/build.make:100: core/unit_test/Kokkos_CoreUnitTest_DeviceAndThreads] Error 1
[CUDA-12.2-NVHPC-AS-HOST-COMPILER] make[1]: *** [CMakeFiles/Makefile2:2335: core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_DeviceAndThreads.dir/all] Error 2
[CUDA-12.2-NVHPC-AS-HOST-COMPILER] make[1]: *** Waiting for unfinished jobs....
```

CUDA-13.0-NVCC-DEBUG
```
[CUDA-13.0-NVCC-DEBUG] [  6%] Linking CXX executable Kokkos_CoreUnitTest_DeviceAndThreads
[CUDA-13.0-NVCC-DEBUG] /usr/bin/ld: ../src/libkokkoscore.so.5.0.99: undefined reference to `cuStreamGetCtx'
[CUDA-13.0-NVCC-DEBUG] /usr/bin/ld: ../src/libkokkoscore.so.5.0.99: undefined reference to `cuCtxGetDevice'
[CUDA-13.0-NVCC-DEBUG] /usr/bin/ld: ../src/libkokkoscore.so.5.0.99: undefined reference to `cuCtxPushCurrent_v2'
[CUDA-13.0-NVCC-DEBUG] collect2: error: ld returned 1 exit status
[CUDA-13.0-NVCC-DEBUG] make[2]: *** [core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_DeviceAndThreads.dir/build.make:101: core/unit_test/Kokkos_CoreUnitTest_DeviceAndThreads] Error 1
[CUDA-13.0-NVCC-DEBUG] make[1]: *** [CMakeFiles/Makefile2:2435: core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_DeviceAndThreads.dir/all] Error 2
[CUDA-13.0-NVCC-DEBUG] make[1]: *** Waiting for unfinished jobs....
```

HIP-ROCm-6.3: connectivity issue to lascaux_02

SYCL-OneAPI: container issue
